src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs

```diff
@@ -174,6 +177,50 @@ internal virtual HttpClient GetHttpClient(bool handleRedirect)
                 handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
                 handler.ClientCertificateOptions = ClientCertificateOption.Manual;
             }
+            else if (CertificateValidationScript != null)
+            {
+                // validationCallBackWrapper wraps the certificateValidationClosure ScriptBlock so the async callback can set the default
+                // PowerShell Runspace in the async thread to the Runspace of the current thread. This provides the ScriptBlock access to 
+                // the current scope.
+                Runspace currentRunspace = Runspace.DefaultRunspace;
+                ScriptBlock certificateValidationClosure = CertificateValidationScript.GetNewClosure();
+                Func<HttpRequestMessage,X509Certificate2,X509Chain,SslPolicyErrors,bool> validationCallBackWrapper = 
+                    delegate(HttpRequestMessage httpRequestMessage, X509Certificate2 x509Certificate2, X509Chain x509Chain, SslPolicyErrors sslPolicyErrors)
+                    {
+                        Runspace previousRunspace = Runspace.DefaultRunspace;
+                        Runspace.DefaultRunspace = currentRunspace;
```

Comment:
This is worrisome. From reading just this code, I have no clue if `currentRunspace` will be blocked while the delegate is running.

For example, if we're streaming objects and in the middle of a call to `WriteObject` in the cmdlet's thread, that could yield to a random PowerShell event handler, which will be a race condition with whatever this delegate might do.

Reply:
I'm not sure how to make this work in a different runspace and supporting access to the current scope. To me, personally, that is not an issue as I can live with this not having scope access. But in the discussion the issue was brought up that users might find that confusing. or difficult to work with.

Can you mock up something that would cause this race condition so I can test it? I did some basic tests with this running in a pipeline and as part of a process block, but I was never able to cause a hang. or unexpected behavior. I thought maybe since we are forcing the call to run synchronously it might nor be an issue:

https://github.com/PowerShell/PowerShell/blob/aea561f08df3cc1c6ce92135db47604868fade2c/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs#L468

But, honestly, I'm not anywhere near an expert on this area. I just know that if I don't specify a run space here I get the exception about a missing runspace.

Reply:
Sure, before running your code, run something like:

```powershell
$t = [System.Timers.Timer]::new(10)
$t.AutoReset = $true
Register-ObjectEvent -InputObject $t -EventName Elapsed -Action { Write-Host 'event' }
$t.Enabled = $true
```

If this doesn't cause problems, try something less trivial in the action - call functions, maybe a recursive function calculating fibonacci or something.

In your certificate validation script, also do something similarly non-trivial, e.g. compute a different fibonacci number.

I'm thinking some state will be corrupted, e.g. the current scope.

Reply:
Using that event and a simple block caused it to hang the thread. So definitely some corruption. ☹️ I'm looking at some other options.

Reply:
10ms might be too fast, maybe try a little slower, say 100ms. The goal was to hit a race as quickly as possible and get something other than a hang, though I suppose a hang is one possible outcome of a race.

Reply:
@lzybkr actually... I think that test may kill any callback, not just the one in this issue.

I reduced the code to this
```csharp
Func<HttpRequestMessage,X509Certificate2,X509Chain,SslPolicyErrors,bool> validationCallBackWrapper = 
    delegate(HttpRequestMessage httpRequestMessage, X509Certificate2 x509Certificate2, X509Chain x509Chain, SslPolicyErrors sslPolicyErrors)
    {
        Boolean result = false;
        try
        {
            result = LanguagePrimitives.IsTrue(
                CertificateValidationScript.InvokeReturnAsIs(
                    httpRequestMessage, 
                    x509Certificate2, 
                    x509Chain, 
                    sslPolicyErrors
                )
            );
        }
        catch // Treat all exceptions as Certificate failures.
        {
            result = false;
        }
        return result;
    };

handler.ServerCertificateCustomValidationCallback = validationCallBackWrapper;
```
No more closure, no more setting the default runspace, no more auto variables or erroraction. Just wrapping the `InvokeReturnAsIs()` in the delegate and this hangs:

```powershell
$t = [System.Timers.Timer]::new(10)
$t.AutoReset = $true
Register-ObjectEvent -InputObject $t -EventName Elapsed -Action { write-host =  'event' }
$t.Enabled = $true
$script = { return $true }
Invoke-RestMethod -uri https://httpbin.org/get -CertificateValidationScript $Script
```

if I reduce it back to 

```csharp
Func<HttpRequestMessage,X509Certificate2,X509Chain,SslPolicyErrors,bool> validationCallBackWrapper =
    LanguagePrimitives.ConvertTo<Func<HttpRequestMessage,X509Certificate2,X509Chain,SslPolicyErrors,bool>>(CertificateValidationScript);
handler.ServerCertificateCustomValidationCallback = validationCallBackWrapper;
```

I get this:

```none
Invoke-RestMethod : There is no Runspace available to run scripts in this thread. You can provide one in the
DefaultRunspace property of the System.Management.Automation.Runspaces.Runspace type. The script block you attempted
to invoke was:  return $true
```
```

Reply:
100ms didn't make a difference. ☹️ Still hangs on all the scenarios.

Reply:
This code has been changed and the setting/unsetting of the runspace is no longer a part of it. However, the callback still results in a thread hang when using the event trigger(s) provided by @lzybkr. I'm not sure where to go on that issue.